src/shared/components/results/ResultsView.tsx

```diff
@@ -224,107 +154,46 @@ const cardRenderer = (
-  sortDirection: SortDirection
-): ColumnDescriptor[] =>
-  columns?.map((columnName) => {
-    const columnConfig = ColumnConfigurations[namespace]?.get(columnName);
-    if (columnConfig) {
-      const columnDescriptor = {
-        label: columnConfig.label,
-        name: columnName,
-        render: (row: APIModel) =>
-          columnConfig.render(convertRow(row, namespace)),
-      };
-      if (sortableColumnToSortColumn.has(columnName)) {
-        return {
-          ...columnDescriptor,
-          sortable: true,
-          sorted: columnName === sortColumn ? sortDirection : undefined,
-        };
-      }
-      return columnDescriptor;
-    }
-    return {
-      label: columnName,
-      name: columnName,
-      render: () => (
-        <div className="warning">{`${columnName} has no config yet`}</div>
-      ),
-    };
-  }) || [];
-
 type ResultsTableProps = {
   selectedEntries: string[];
   handleEntrySelection: (rowId: string) => void;
-  sortableColumnToSortColumn: Map<Column, string>;
 };
 
 const ResultsView: FC<ResultsTableProps> = ({
   selectedEntries,
   handleEntrySelection,
-  sortableColumnToSortColumn,
 }) => {
   const namespace = useNS() || Namespace.uniprotkb;
   const [viewMode] = useUserPreferences<ViewMode>('view-mode', ViewMode.CARD);
-  const [columns] = useUserPreferences<Column[]>(
-    `table columns for ${namespace}` as const,
-    nsToDefaultColumns[namespace]
-  );
+  const history = useHistory();
+  const [columns, updateColumnSort] = useColumns();
 
   const prevNamespace = useRef<Namespace>(namespace);
   useEffect(() => {
     // will set it *after* the current render
     prevNamespace.current = namespace;
   });
-  const prevColumns = useRef<Column[] | undefined>(columns);
+  const prevColumns = useRef<ColumnDescriptor[] | undefined>(columns);
   useEffect(() => {
     // will set it *after* the current render
     prevColumns.current = columns;
   });
 
-  const history = useHistory();
-  const location = useLocation();
-
-  const { search: queryParamFromUrl } = location;
-  const {
-    query,
-    selectedFacets,
-    sortColumn,
-    sortDirection,
-    direct,
-  } = getParamsFromURL(queryParamFromUrl);
-
-  let queryColumns = viewMode === ViewMode.CARD ? undefined : columns;
+  let queryColumns =
+    viewMode === ViewMode.CARD
+      ? undefined
+      : columns.map((columnDescriptor) => columnDescriptor.name as Column);
   if (viewMode === ViewMode.CARD) {
     // TODO: Do similar things for the rest of namespaces
     if (namespace === Namespace.uniprotkb) {
       queryColumns = fieldsForUniProtKBCards;
     }
   }
 
-  const initialApiUrl = getAPIQueryUrl({
-    namespace,
-    query,
-    // TODO: Do similar things for the rest of namespaces
-    columns: queryColumns,
-    selectedFacets,
-    // Not really interested in the facets here, so try to reduce payload
-    facets: null,
-    sortColumn,
-    sortDirection,
+  const { url: initialApiUrl, direct } = useNSQuery({
```

Comment:
When switching from table to card view initialApiUrl is updated however it's not getting into this `useEffect`:
```
  useEffect(() => {
    setAllResults([]);
    setMetaData({ total: 0, nextUrl: undefined });
    setUrl(initialApiUrl);
  }, [initialApiUrl]);
```
to reset and fetch all of the columns needed for the card view.

Reply:
I've simplified the "hacky logic" to force a re-render of the loader, and I think this is now addressed. @aurel-l could you have a look if you don't mind as I think you'd added this logic in the first place?

Reply:
Using the initial URL to detect when to force a refresh is good, you should also be able to drop the `prevViewMode.current !== viewMode` bit too I believe.

Reply:
I tried that and there's a race condition and the `useEffect` in `useNSQuery.ts` hasn't run by the time it gets to the check. Not sure there's an easy way around that